File: python/pascals-triangle/pascals_triangle.py
triangle = []
def rows(row_count):
    global triangle
    triangle = []
    if row_count < 0: raise ValueError("number of rows is negative")
    pascal(row_count)
    return triangle

def pascal(rows_left):
    global triangle
    if rows_left == 0: return
    if not triangle:
        triangle.append([1])
    else:
        prev = triangle[-1]
        cur = []
        for i in range(len(prev) + 1):
            before = prev[i - 1] if i > 0 else 0
            after = prev[i] if i < len(prev) else 0
            cur.append(before + after)
        triangle.append(cur)
    return pascal(rows_left - 1)

File: ruby/pascals-triangle/pascals_triangle.rb
class Triangle
  def initialize(size)
    @size = size
  end

  def rows
    (1..@size).map.with_object([]) { |y, lines|
      lines << (1..y).map { |x|
        ((x == 1 or x == y) ? 1 : lines[y-2][x-1] + lines[y-2][x-2])
      }
    }
  end
end
